### ISSUE TITLE
No Content (204) Responses are non functional

### ISSUE THREAD

**Issue by ptravers** [twilio/guardrail#121]
The client requires the No Content response to have content due to the usage of `isSuccessful` to determine whether to consume the entity. The server also requires that a No Content response not contain any content and converts it in to a 500 Internal Server Error if it does have content. This results in a Catch 22 that is not avoidable in a sane way.

**Comment by blast-hardcheese** [comment]
This is definitely undesirable. Most of the APIs I've been integrating with have had one "success" status code, so an expedient was to treat them all as having the same schema, focusing on the server generator for the strongly typed code.

The strategy for implementing what you're looking for here would involve some pretty drastic changes from the user's perspective. For example, a new interface could look like:

```scala
FooClient.getBar(): Either[Throwable, getBarResponse]
```

with `getBarResponse` being used as:
```scala
FooClient.getBar().fold(???, _.fold(response => {
  case response.Ok(successValue) => ???
  case response.NoContent => ???
}))
```

where `response` only contains the valid response types from the downstream service. This would have the unfortunate side-effect of making the type of `getBar` very complex, so hiding that complexity in a `FooClient.getBarResponse` type alias could make it easier for consumers to implement around.

This is the direction I'm leaning, but I'd be interested on your thoughts as a consumer of this interface. Any alternative to this approach would also need to raise a compiler error in the case of the specification changing out from under an implementation.

**Comment by ptravers** [comment]
Yeah so I would be curious as to a where in the codebase the client is generated and as to whether or not it would not just be easier to replace a line in  
```
private[this] def wrap[T: FromEntityUnmarshaller](client: HttpClient, request: HttpRequest): EitherT[Future, Either[Throwable, HttpResponse], T] = {
    EitherT(client(request).flatMap(resp => if (resp.status.isSuccess) {
      Unmarshal(resp.entity).to[T].map(Right.apply _)
    } else {
      FastFuture.successful(Left(Right(resp)))
    }).recover({
      case e: Throwable =>
        Left(Left(e))
    }))
  }
```
if instead `if (resp.status.isSuccess && resp.status != NoContent)` was used surely that would avoid this issue?

the server is probably correct in this instance but maybe also overly zealous.

sorry for the slow response

**Comment by ptravers** [comment]
I tracked it down. https://github.com/twilio/guardrail/blob/28c4aad69be29485688fad951cb8b7c0434535b5/modules/codegen/src/main/scala/com/twilio/guardrail/generators/AkkaHttpClientGenerator.scala#L413

do you see any reason this could not be altered in the way described above? I am happy to make the change and tests for it.

**Comment by blast-hardcheese** [comment]
`wrap` is defined here: https://github.com/twilio/guardrail/blob/49fe1082b5b6d7979723cbb58ebcb22e4a90268b/modules/codegen/src/main/scala/com/twilio/guardrail/generators/AkkaHttpClientGenerator.scala#L410

To be clear, your server specification defines something similar to the following, with at least one other `2xx` response than 204, correct?

```yaml
        responses:
          200:
            schema:
              "$ref": "#/definitions/Foo"
          204:
            description: "No content"
```

If this is not the case, (if you only have a single 2xx response, which is the aforementioned 204,) then your solution would be a good stopgap for the time being, and I'd be interested in releasing that as a bugfix. I was thinking about the additional case of:

```yaml
        responses:
          200:
            schema:
              "$ref": "#/definitions/Foo"
          201:
            schema:
              "$ref": "#/definitions/Bar"
          202:
            schema:
              "$ref": "#/definitions/Baz"
          204:
            description: "No content"
```

where 201, 202, and 204 all attempt to be decoded as `Foo` due to [`getBestSuccessResponse`](https://github.com/twilio/guardrail/blob/49fe1082b5b6d7979723cbb58ebcb22e4a90268b/modules/codegen/src/main/scala/com/twilio/guardrail/SwaggerUtil.scala#L348-L351).

**Comment by blast-hardcheese** [comment]
Totally missed https://github.com/twilio/guardrail/issues/121#issuecomment-435109316 . Yeah, as for the stopgap, go for it. So long as it's tested, I'm all for it. Note that everything in [`modules/sample/src/test/scala`](https://github.com/twilio/guardrail/tree/master/modules/sample/src/test/scala) are tests on the _generated_ code, so not only can we guarantee that the 204 typechecks, but we can actually verify that 204s round-trip as well.

**Comment by blast-hardcheese** [comment]
This is resolved by #152

**Comment by blast-hardcheese** [comment]
Feel free to reopen if necessary
